<a href="https://colab.research.google.com/github/gaurav873/Gen--AI/blob/main/RAG_Pipeline.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import os

In [ ]:
from dotenv import load_dotenv

In [ ]:
load_dotenv()

True

![image.png](attachment:image.png)

In [ ]:
GOOGLE_API_KEY = os.getenv("GOOGLE_API_KEY")

In [ ]:
GROQ_API_KEY=os.getenv("GROQ_API_KEY")

In [ ]:
os.environ["GOOGLE_API_KEY"] = GOOGLE_API_KEY

In [ ]:
os.environ["GROQ_API_KEY"] = GROQ_API_KEY

In [ ]:
from langchain_google_genai import GoogleGenerativeAIEmbeddings
embeddings = GoogleGenerativeAIEmbeddings(model="gemini-embedding-001")

d:\complete_content_new\Full-Stack-GenAI-Bootcamp-1.0\env\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [ ]:
os.environ["OPENAI_API_KEY"] = os.getenv("OPENAI_API_KEY")

In [ ]:
embeddings.embed_query("Hello world")

[-0.02342152,
 0.01676572,
 0.009261323,
 -0.06383,
 -0.0026262768,
 0.0010187156,
 -0.01125684,
 0.0130036585,
 0.008751671,
 0.0012745215,
 -0.0007880878,
 -0.019086141,
 0.030971918,
 0.044264916,
 0.11137619,
 0.018025596,
 -0.0031383373,
 -0.0130702155,
 0.0054121013,
 -0.009036983,
 0.009904047,
 -0.010718052,
 0.012551899,
 0.011420718,
 -0.03375867,
 0.008877066,
 0.027795823,
 0.0037452083,
 0.029841818,
 0.019861836,
 0.018075233,
 -0.007178086,
 -0.02497957,
 0.011720823,
 -0.0073324037,
 0.0068515264,
 0.011618152,
 0.004278904,
 -0.0099426275,
 -0.009696086,
 -0.0037028084,
 0.006349137,
 -0.012098703,
 -0.015629271,
 -0.022799378,
 -0.012305125,
 -0.01497548,
 -0.006503783,
 -0.005624279,
 0.020265369,
 -0.029982245,
 -0.008486281,
 -0.0050998786,
 -0.14985937,
 -0.017222198,
 0.011674307,
 -0.014542328,
 0.02246327,
 -0.009126675,
 -0.018621653,
 -0.004780002,
 0.0020153068,
 -0.0133157885,
 -0.006439805,
 0.0008783784,
 -0.023549458,
 -0.008333907,
 0.01863065,
 -0.0190

In [ ]:
from langchain_openai import OpenAIEmbeddings

In [ ]:
embeddings=OpenAIEmbeddings(model="text-embedding-3-large")

In [ ]:
embeddings=OpenAIEmbeddings(model="text-embedding-3-small")

In [ ]:
embeddings.embed_query("Hello world")

[-0.0021610260009765625,
 -0.049072265625,
 0.0209808349609375,
 0.0313720703125,
 -0.045318603515625,
 -0.0264129638671875,
 -0.028961181640625,
 0.060302734375,
 -0.0257110595703125,
 -0.0148162841796875,
 0.015472412109375,
 -0.0300445556640625,
 -0.0203704833984375,
 -0.033416748046875,
 0.0258331298828125,
 0.01422119140625,
 -0.070068359375,
 0.0123748779296875,
 0.01483917236328125,
 0.048919677734375,
 0.020782470703125,
 -0.00887298583984375,
 -0.0151214599609375,
 -0.016632080078125,
 0.0259552001953125,
 -0.002857208251953125,
 -0.024383544921875,
 0.0242919921875,
 0.0018329620361328125,
 -0.05572509765625,
 0.023101806640625,
 -0.04547119140625,
 -0.00870513916015625,
 0.0030956268310546875,
 0.004558563232421875,
 0.0018224716186523438,
 0.0266876220703125,
 0.01013946533203125,
 -0.01201629638671875,
 -0.0115203857421875,
 -0.01491546630859375,
 -0.0231475830078125,
 0.025390625,
 0.03680419921875,
 -0.035491943359375,
 0.02130126953125,
 -0.0631103515625,
 0.04037475585

In [ ]:
from langchain_google_genai import ChatGoogleGenerativeAI
llm = ChatGoogleGenerativeAI(model="gemini-3.1-flash-lite")

In [ ]:
llm.invoke("hi how are you?").content[0]["text"]

"I'm doing great, thank you for asking! How are you doing today? Is there anything I can help you with?"

#### It is simple AI Assistant without RAG Pipelie

In [ ]:
while True:
    user_input = input("You: ")
    if user_input.lower() in ["exit", "quit"]:
        print("Exiting the chat. Goodbye!")
        break
    response = llm.invoke(user_input).content[0]["text"]
    print(f"AI: {response}")

AI: As of right now, **there is no such model as "DeepSeek-V4-Pro."**

It is likely you are conflating several different naming conventions. Here is the current status of the DeepSeek model lineup:

1.  **DeepSeek-V3:** This is the latest flagship model released by DeepSeek. It is a Mixture-of-Experts (MoE) model with **671 billion parameters** in total, of which approximately 37 billion are active per token.
2.  **DeepSeek-R1:** This is the latest reasoning-focused model. It is based on the DeepSeek-V3 architecture and shares the same scale (671 billion parameters total).
3.  **Future Versions:** DeepSeek has not announced a "V4" or any "Pro" version of their models.

If you are seeing references to a "DeepSeek-V4-Pro" online, it is almost certainly a hallucination, a misunderstanding of the current version numbers, or misinformation. 

**Summary of the current flagship (DeepSeek-V3/R1):**
*   **Total Parameters:** 671 Billion
*   **Active Parameters:** ~37 Billion
*   **Architecture:

In [ ]:
from langchain_community.document_loaders import TextLoader, DirectoryLoader

C:\Users\Sunny\AppData\Local\Temp\ipykernel_18312\23717305.py:1: DeprecationWarning: `langchain-community` is being sunset and is no longer actively maintained. See https://github.com/langchain-ai/langchain-community/issues/674 for details and migration guidance toward standalone integration packages.
  from langchain_community.document_loaders import TextLoader, DirectoryLoader


In [ ]:
from langchain_community.document_loaders import WebBaseLoader

USER_AGENT environment variable not set, consider setting it to identify your requests.


In [ ]:
from langchain_community.document_loaders import PyPDFLoader

In [ ]:
file_path = "D:\complete_content_new\Full-Stack-GenAI-Bootcamp-1.0\Class-31-11-July-2026\deepseek-v4-2026.pdf"

In [ ]:
pages = PyPDFLoader(file_path).load()

In [ ]:
len(pages)

58

In [ ]:
pages

[Document(metadata={'producer': 'pikepdf 8.15.1', 'creator': 'arXiv GenPDF (tex2pdf:a6404ea)', 'creationdate': '', 'author': 'DeepSeek-AI; Anyi Xu; Bangcai Lin; Bing Xue; Bingxuan Wang; Bingzheng Xu; Bochao Wu; Bowei Zhang; Chaofan Lin; Chen Dong; Chenchen Ling; Chengda Lu; Chenggang Zhao; Chengqi Deng; Chengyu Hou; Chenhao Xu; Chenze Shao; Chong Ruan; Conner Sun; Damai Dai; Daya Guo; Dejian Yang; Deli Chen; Donghao Li; Dongjie Ji; Erhang Li; Fang Wei; Fangyun Lin; Fangzhou Yuan; Feiyu Xia; Fucong Dai; Guangbo Hao; Guanting Chen; Guoai Cao; Guolai Meng; Guowei Li; Han Yu; Han Zhang; Hanwei Xu; Hao Li; Haofen Liang; Haoling Zhang; Haoming Luo; Haoran Wei; Haotian Yuan; Haowei Zhang; Haowen Luo; Haoyu Chen; Haozhe Ji; Hengqing Zhang; Honghui Ding; Hongxuan Tang; Huanqi Cao; Huazuo Gao; Hui Qu; Hui Zeng; J Yang; JQ Zhu; Jia Luo; Jia Song; Jia Yu; Jialiang Huang; Jialu Cai; Jian Liang; Jiangting Zhou; Jiasheng Ye; Jiashi Li; Jiaxin Xu; Jiewen Hu; Jieyu Yang; Jin Chen; Jin Yan; Jingchang Ch

In [ ]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

In [ ]:
#parameter values will discuss in the upcoming session
chunker =RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=400)

In [ ]:
chunked_data = chunker.split_documents(pages)

In [ ]:
len(chunked_data)

283

In [ ]:
chunked_data

[Document(metadata={'producer': 'pikepdf 8.15.1', 'creator': 'arXiv GenPDF (tex2pdf:a6404ea)', 'creationdate': '', 'author': 'DeepSeek-AI; Anyi Xu; Bangcai Lin; Bing Xue; Bingxuan Wang; Bingzheng Xu; Bochao Wu; Bowei Zhang; Chaofan Lin; Chen Dong; Chenchen Ling; Chengda Lu; Chenggang Zhao; Chengqi Deng; Chengyu Hou; Chenhao Xu; Chenze Shao; Chong Ruan; Conner Sun; Damai Dai; Daya Guo; Dejian Yang; Deli Chen; Donghao Li; Dongjie Ji; Erhang Li; Fang Wei; Fangyun Lin; Fangzhou Yuan; Feiyu Xia; Fucong Dai; Guangbo Hao; Guanting Chen; Guoai Cao; Guolai Meng; Guowei Li; Han Yu; Han Zhang; Hanwei Xu; Hao Li; Haofen Liang; Haoling Zhang; Haoming Luo; Haoran Wei; Haotian Yuan; Haowei Zhang; Haowen Luo; Haoyu Chen; Haozhe Ji; Hengqing Zhang; Honghui Ding; Hongxuan Tang; Huanqi Cao; Huazuo Gao; Hui Qu; Hui Zeng; J Yang; JQ Zhu; Jia Luo; Jia Song; Jia Yu; Jialiang Huang; Jialu Cai; Jian Liang; Jiangting Zhou; Jiasheng Ye; Jiashi Li; Jiaxin Xu; Jiewen Hu; Jieyu Yang; Jin Chen; Jin Yan; Jingchang Ch

In [ ]:
chunked_data[0]

Document(metadata={'producer': 'pikepdf 8.15.1', 'creator': 'arXiv GenPDF (tex2pdf:a6404ea)', 'creationdate': '', 'author': 'DeepSeek-AI; Anyi Xu; Bangcai Lin; Bing Xue; Bingxuan Wang; Bingzheng Xu; Bochao Wu; Bowei Zhang; Chaofan Lin; Chen Dong; Chenchen Ling; Chengda Lu; Chenggang Zhao; Chengqi Deng; Chengyu Hou; Chenhao Xu; Chenze Shao; Chong Ruan; Conner Sun; Damai Dai; Daya Guo; Dejian Yang; Deli Chen; Donghao Li; Dongjie Ji; Erhang Li; Fang Wei; Fangyun Lin; Fangzhou Yuan; Feiyu Xia; Fucong Dai; Guangbo Hao; Guanting Chen; Guoai Cao; Guolai Meng; Guowei Li; Han Yu; Han Zhang; Hanwei Xu; Hao Li; Haofen Liang; Haoling Zhang; Haoming Luo; Haoran Wei; Haotian Yuan; Haowei Zhang; Haowen Luo; Haoyu Chen; Haozhe Ji; Hengqing Zhang; Honghui Ding; Hongxuan Tang; Huanqi Cao; Huazuo Gao; Hui Qu; Hui Zeng; J Yang; JQ Zhu; Jia Luo; Jia Song; Jia Yu; Jialiang Huang; Jialu Cai; Jian Liang; Jiangting Zhou; Jiasheng Ye; Jiashi Li; Jiaxin Xu; Jiewen Hu; Jieyu Yang; Jin Chen; Jin Yan; Jingchang Che

In [ ]:
chunked_data[0].metadata

{'producer': 'pikepdf 8.15.1',
 'creator': 'arXiv GenPDF (tex2pdf:a6404ea)',
 'creationdate': '',
 'author': 'DeepSeek-AI; Anyi Xu; Bangcai Lin; Bing Xue; Bingxuan Wang; Bingzheng Xu; Bochao Wu; Bowei Zhang; Chaofan Lin; Chen Dong; Chenchen Ling; Chengda Lu; Chenggang Zhao; Chengqi Deng; Chengyu Hou; Chenhao Xu; Chenze Shao; Chong Ruan; Conner Sun; Damai Dai; Daya Guo; Dejian Yang; Deli Chen; Donghao Li; Dongjie Ji; Erhang Li; Fang Wei; Fangyun Lin; Fangzhou Yuan; Feiyu Xia; Fucong Dai; Guangbo Hao; Guanting Chen; Guoai Cao; Guolai Meng; Guowei Li; Han Yu; Han Zhang; Hanwei Xu; Hao Li; Haofen Liang; Haoling Zhang; Haoming Luo; Haoran Wei; Haotian Yuan; Haowei Zhang; Haowen Luo; Haoyu Chen; Haozhe Ji; Hengqing Zhang; Honghui Ding; Hongxuan Tang; Huanqi Cao; Huazuo Gao; Hui Qu; Hui Zeng; J Yang; JQ Zhu; Jia Luo; Jia Song; Jia Yu; Jialiang Huang; Jialu Cai; Jian Liang; Jiangting Zhou; Jiasheng Ye; Jiashi Li; Jiaxin Xu; Jiewen Hu; Jieyu Yang; Jin Chen; Jin Yan; Jingchang Chen; Jingli Zhou;

In [ ]:
chunked_data[0].page_content

'DeepSeek-V4:\nTowards Highly Efficient Million-Token Context Intelligence\nDeepSeek-AI\nresearch@deepseek.com\nAbstract\nWe present a preview version of DeepSeek-V4 series, including two strong Mixture-of-\nExperts (MoE) language models — DeepSeek-V4-Pro with 1.6T parameters (49B activated) and\nDeepSeek-V4-Flash with 284B parameters (13B activated) — both supporting a context length of\none million tokens. DeepSeek-V4 series incorporate several key upgrades in architecture and op-\ntimization: (1) a hybrid attention architecture that combines Compressed Sparse Attention (CSA)\nand Heavily Compressed Attention (HCA) to improve long-context efficiency; (2) Manifold-\nConstrained Hyper-Connections (mHC) that enhance conventional residual connections; (3)\nand the Muon optimizer for faster convergence and greater training stability. We pre-train\nboth models on more than 32T diverse and high-quality tokens, followed by a comprehensive'

In [ ]:
len(embeddings.embed_query("Hello world"))

1536

In [ ]:
## will take one specific session for the vector database
import faiss
from langchain_community.vectorstores import FAISS
from langchain_community.docstore.in_memory import InMemoryDocstore

In [ ]:
index = faiss.IndexFlatL2(1536)

In [ ]:
vector_store = FAISS(
    embedding_function=embeddings,
    index=index,
    docstore=InMemoryDocstore(),
    index_to_docstore_id={}
)

In [ ]:
vector_store.add_documents(chunked_data)

['87e9fcfa-b93b-4e33-bad7-eff17906a721',
 'b89cd453-297d-4d9b-bb2e-816e944c45af',
 'c28bd76f-4d54-4f9e-bbcb-9a307551c595',
 '03f2b79d-b5b0-4b65-867f-2a8504bbf089',
 '8b7c43ff-7793-478c-b3ed-27911cd4ef27',
 '26c48653-e90d-4607-846e-699041c4fe46',
 '1ef755e0-6073-4739-8c60-0c068eaa9df3',
 'aaa0d3c1-1490-4c51-9d1f-e316c8c00d5c',
 '83788fbb-3cf6-4a6e-9bf0-d913d814b629',
 'dcdf9d99-6489-444c-a141-ab4c8e65e19f',
 'aedf446a-ffc3-4051-8b16-37ca1f23632f',
 '4d532895-a29f-497c-8f55-d9e082bba702',
 '6be3bae9-e70d-4e44-b1d8-3bca8a79ee6f',
 '7e1233e3-f1ad-4a4f-ab2d-8cae353f6ac7',
 '9c426363-bfcc-4cc4-b91d-94336ddf3b29',
 'd0af37c7-d986-4057-8709-ffdc544922b5',
 '61d2a825-1e1f-460c-bcb3-5f1544de7333',
 'fbbf7702-5e74-40bd-b5ed-9a7ed99fcd7d',
 '6d3d731d-8d44-427c-a155-802f328b4194',
 'e48e3f52-5fa9-47db-ab16-40a712d2eeeb',
 '083232b7-be5f-4d9a-b99a-cd6384e1d0d8',
 'b89ac128-fc85-4daa-ac58-a72f5616b630',
 '98cce679-45d7-4ef7-bebd-24f7345bd2c0',
 'ae883cc3-c2f2-4a1c-8283-40c40c8cb1c7',
 '9fa09a20-f2ee-

In [ ]:
retriever = vector_store.as_retriever(search_kwargs={"k": 10})

In [ ]:
retriever.invoke("what is this deepseek-v4-2026 model about?")

[Document(id='9ee241b5-5114-4002-86fc-07516dbd364b', metadata={'producer': 'pikepdf 8.15.1', 'creator': 'arXiv GenPDF (tex2pdf:a6404ea)', 'creationdate': '', 'author': 'DeepSeek-AI; Anyi Xu; Bangcai Lin; Bing Xue; Bingxuan Wang; Bingzheng Xu; Bochao Wu; Bowei Zhang; Chaofan Lin; Chen Dong; Chenchen Ling; Chengda Lu; Chenggang Zhao; Chengqi Deng; Chengyu Hou; Chenhao Xu; Chenze Shao; Chong Ruan; Conner Sun; Damai Dai; Daya Guo; Dejian Yang; Deli Chen; Donghao Li; Dongjie Ji; Erhang Li; Fang Wei; Fangyun Lin; Fangzhou Yuan; Feiyu Xia; Fucong Dai; Guangbo Hao; Guanting Chen; Guoai Cao; Guolai Meng; Guowei Li; Han Yu; Han Zhang; Hanwei Xu; Hao Li; Haofen Liang; Haoling Zhang; Haoming Luo; Haoran Wei; Haotian Yuan; Haowei Zhang; Haowen Luo; Haoyu Chen; Haozhe Ji; Hengqing Zhang; Honghui Ding; Hongxuan Tang; Huanqi Cao; Huazuo Gao; Hui Qu; Hui Zeng; J Yang; JQ Zhu; Jia Luo; Jia Song; Jia Yu; Jialiang Huang; Jialu Cai; Jian Liang; Jiangting Zhou; Jiasheng Ye; Jiashi Li; Jiaxin Xu; Jiewen Hu; 

In [ ]:
retriever.invoke("what is a release date of deepseek v4 research paper?")[0].page_content

'Figure 10 presents a comparison of performance and cost among DeepSeek-V4-Pro, DeepSeek-\nV4-Flash, and DeepSeek-V3.2 on representative reasoning and agentic tasks. By scaling test-time\ncompute, DeepSeek-V4 series achieve substantial improvements over the predecessor. Further-\nmore, on reasoning tasks like HLE, DeepSeek-V4-Pro demonstrates higher token efficiency than\nDeepSeek-V3.2.\n40'

In [ ]:
retriever.invoke("what is a release date of deepseek v4 research paper?")[1].page_content

'performance highs among DeepSeek base models on the most demanding benchmarks. On\nknowledge-intensive evaluations, it delivers dramatic gains, while also substantially advancing\nlong-context understanding. On most reasoning and code benchmarks, DeepSeek-V4-Pro-Base\nalso exceeds both previous models. This comprehensive uplift confirms DeepSeek-V4-Pro-Base\nas the strongest foundation model in the DeepSeek series, outperforming its predecessors across\nthe spectrum of knowledge, reasoning, coding, and long-context capabilities.\n27'

In [ ]:
retriever.invoke("what is a release date of deepseek v4 research paper?")[2].page_content

'series, including the preview versions of DeepSeek-V4-Pro with 1.6T parameters (49B activated)\nand DeepSeek-V4-Flash with 284B parameters (13B activated). Through architectural innova-\ntions, DeepSeek-V4 series achieve a dramatic leap in computational efficiency for processing\nultra-long sequences. This breakthrough enables efficient support for a context length of one\nmillion tokens, ushering in a new era of million-length contexts for next-generation LLMs. We\nbelieve our capability to efficiently handle ultra-long sequences unlocks the next frontier of\ntest-time scaling, paves the way for deeper research into long-horizon tasks, and establishes a\nnecessary foundation for exploring future paradigms like online learning.\nCompared with the DeepSeek-V3 architecture (DeepSeek-AI, 2024), DeepSeek-V4 series\nretain the DeepSeekMoE framework (Dai et al., 2024) and Multi-Token Prediction (MTP) strategy,'

In [ ]:
template = """Answer the question based only on the following context:
{context}

Question: {question}
"""

In [ ]:
from langchain_core.prompts import PromptTemplate

In [ ]:
prompt=PromptTemplate.from_template(template)

In [ ]:
prompt

PromptTemplate(input_variables=['context', 'question'], input_types={}, partial_variables={}, template='Answer the question based only on the following context:\n{context}\n\nQuestion: {question}\n')

PromptTemplate(input_variables=['context', 'question'], input_types={}, partial_variables={}, template='Answer the question based only on the following context:\n{context}\n\nQuestion: {question}\n')

In [ ]:
def format_docs(docs):
    return "\n\n".join([doc.page_content for doc in docs])

In [ ]:
from langchain_core.output_parsers import StrOutputParser

In [ ]:
from langchain_core.runnables import RunnablePassthrough

In [ ]:
rag_chain = {"context": retriever | format_docs, "question": RunnablePassthrough()} | prompt | llm | StrOutputParser()

Answer from the simple AI Assistant:

AI: As of right now, **DeepSeek has not released a V4 research paper, nor has they officially released a model under the name "DeepSeek V4."**

Here is the current status of the DeepSeek model lineup:

*   **DeepSeek-V3:** This is the latest major model released by DeepSeek. The technical report for DeepSeek-V3 was released on **December 26, 2024**.
*   **DeepSeek-R1:** This is their latest reasoning-focused model (based on the V3 architecture). The paper for DeepSeek-R1 was released on **January 22, 2025**.

Because DeepSeek-V3 and R1 are still very new, there have been no credible announcements or documentation regarding a "DeepSeek-V4." If you have seen mentions of a V4, they are likely speculative or inaccurate rumors.

You can always track their official releases and technical reports directly on their **[official GitHub repository](https://github.com/deepseek-ai)** or their [research website](https://api.deepseek.com/).
Exiting the chat. Goodbye!

In [ ]:
"DeepSeek-V4-Pro with 1.6T parameters"

In [ ]:
rag_chain.invoke("what is a release date of DeepSeek-V4 research paper?")

'Based on the provided text, there is no information regarding the release date of the DeepSeek-V4 research paper.'

In [ ]:
rag_chain.invoke("how many parameters are there in the DeepSeek-V4-Pro model?")

'Based on the provided text, the DeepSeek-V4-Pro model comprises 1.6T total parameters, of which 49B are activated for each token.'

I: As of right now, **there is no such model as "DeepSeek-V4-Pro."**

It is likely you are conflating several different naming conventions. Here is the current status of the DeepSeek model lineup:

1.  **DeepSeek-V3:** This is the latest flagship model released by DeepSeek. It is a Mixture-of-Experts (MoE) model with **671 billion parameters** in total, of which approximately 37 billion are active per token.
2.  **DeepSeek-R1:** This is the latest reasoning-focused model. It is based on the DeepSeek-V3 architecture and shares the same scale (671 billion parameters total).
3.  **Future Versions:** DeepSeek has not announced a "V4" or any "Pro" version of their models.

If you are seeing references to a "DeepSeek-V4-Pro" online, it is almost certainly a hallucination, a misunderstanding of the current version numbers, or misinformation.

**Summary of the current flagship (DeepSeek-V3/R1):**
*   **Total Parameters:** 671 Billion
*   **Active Parameters:** ~37 Billion
*   **Architecture:** Mixture-of-Experts (MoE)

In [ ]:
"DeepSeek-V4-Flash"

In [ ]:
rag_chain.invoke("how many parameters are there in the DeepSeek-V4-Flash model?")

'Based on the provided text, the DeepSeek-V4-Flash model has 284 billion total parameters (with 13 billion activated parameters).'

In [ ]:
rag_chain.invoke("which Reinforcement Learning is used in the DeepSeek-V4 model?")

'Based on the provided text, the DeepSeek-V4 series uses Group Relative Policy Optimization (GRPO) (DeepSeek-AI, 2025) as the reinforcement learning method during its post-training pipeline.'

'Based on the provided text, the DeepSeek-V4 series uses Group Relative Policy Optimization (GRPO) (DeepSeek-AI, 2025) as the reinforcement learning method during its post-training pipeline.'